src/std/text/csv.ss

```diff
@@ -0,0 +1,332 @@
+       (csv-eol +lf+)
+       (csv-line-endings [+crlf+ +lf+])
+       (csv-skip-whitespace? #f))
+    (thunk)))
+
+(def (call-with-strict-rfc4180-csv-syntax thunk)
+  (call-with-rfc4180-csv-syntax
+   (lambda () (parameterize ((csv-line-endings '(+crlf+)) (csv-allow-binary? #f))
+                (thunk)))))
+
+(def (valid-eol? x)
+  (member x [+cr+ +lf+ +crlf+]))
+
+(def (validate-csv-parameters)
+  (assert! (char? (csv-separator)))
+  (assert! (char? (csv-quote)))
+  (assert! (not (equal? (csv-separator) (csv-quote))))
+  (assert! (boolean? (csv-unquoted-quotequote?)))
+  (assert! (boolean? (csv-loose-quote?)))
+  (assert! (boolean? (csv-skip-whitespace?)))
+  (assert! (valid-eol? (csv-eol)))
+  (assert! (not (member (string-ref (csv-eol) 0) [(csv-separator) (csv-quote)])))
+  (assert! (and (list? (csv-line-endings)) (every valid-eol? (csv-line-endings)))))
+
+;; For internal use only
+(def csv-eol-matrix (make-parameter #(#t #t #t))) ;; internal: do we accept cr? lf? crlf?
+
+(def (csv-eol-matrix<-line-endings line-endings)
+  (def (? x) (and (member x line-endings) #t))
+  (vector (? +cr+) (? +lf+) (? +crlf+)))
+
+; -----------------------------------------------------------------------------
+;;; The parser
+
+;; Is character c some kind of white space?
+;; NB: this only handles a tiny subset of whitespace characters, even if restricted to ASCII.
+;; However, it's rather portable, and it is what the creativyst document specifies.
+;; Be careful to not skip a separator, as it could be e.g. a tab!
+(def (char-csv-space? c)
+  (and (or (eqv? c #\space) (eqv? c #\tab)) (not (eqv? c (csv-separator)))))
+
+(def (char-ascii-text? c)
+  (let ((i (char->integer c)))
+    (or (<= #x20 i #x7E) (= i 10) (= i 13))))
+
+(def (accept pred port)
+  (and (pred (peek-char port)) (read-char port)))
+
+(def (accept-char c port)
+  (accept (cut eqv? c <>) port))
+
+(def (accept-eof port)
+  (accept eof-object? port))
+
+(def (accept-eol port)
+  (match (csv-eol-matrix)
+    ((vector lf? cr? crlf?)
+     (or (and lf? (accept-char #\newline port) #t)
+         (and (or cr? crlf?)
+              (and (accept-char #\return port)
+                   (or (and crlf? (accept-char #\newline port))
+                       cr?
+                       (raise-io-error 'accept-eol "Carriage-Return without Linefeed!"))))))))
+
+(def (accept-space port)
+  (accept char-csv-space? port))
+
+(def (accept-spaces port)
+  (call-with-list-builder
+   (lambda (c _) (let loop () (let ((x (accept-space port))) (when x (c x) (loop)))))))
+
+(def (accept-quote port)
+  (accept (cut eqv? <> (csv-quote)) port))
+
+(def (accept-separator port)
+  (accept (cut eqv? <> (csv-separator)) port))
+
+;; Read one line from PORT in CSV format, using the current syntax parameters.
+;; Return a list of strings, one for each field in the line.
+;; Entries are read as strings;
+;; it is up to you to interpret the strings as whatever you want.
+(def (read-csv-line port)
+  (validate-csv-parameters)
```

Comment:
that's an expensive check for a single line read step.
I think it shouldn't be here or at the very least offer an unvalidated single line reader.

Reply:
What about moving to read-csv-lines?